src/ros_filter.cpp

```diff
@@ -950,17 +951,23 @@ void RosFilter::loadParams()
     ss << "odom" << topic_ind++;
     std::string odom_topic_name = ss.str();
     std::string odom_topic;
-    more_params = node_->get_parameter(odom_topic_name, odom_topic);
+    node_->declare_parameter(odom_topic_name);
+
+    rclcpp::Parameter parameter;
+    if (rclcpp::PARAMETER_NOT_SET != node_->get_parameter(odom_topic_name, parameter)) {
+      more_params = true;
+      odom_topic = node_->get_parameter(odom_topic_name, parameter);
+    } else {
+        node_->undeclare_parameter(odom_topic_name);
```

Comment:
I feel like we're missing a `more_params = false;` statement here. The previous code would switch the state of `more_params`, but in this loop, once `more_params` is set to _true_, I don't see how it gets set back to false. If I'm right, you'll want to double-check the others.

Reply:
I'll take an extra looks, thanks for your comments, really appreciate it.

Reply:
It is set to `false` again after each `do - while`-statement before the next round,  so regardless which statement is hit here, it will be set to false before it is used again (see line 1087).

Reply:
Right, but my concern is that we'll never break _out_ of this do-while. For example, we set `more_params` to false before the loop here:

https://github.com/cra-ros-pkg/robot_localization/blob/70cd3dd54dc017b0a9167f85ce92f1711df7b0b9/src/ros_filter.cpp#L944

Then, if `odom0` is defined, we set `more_params` to true here:

https://github.com/cra-ros-pkg/robot_localization/blob/70cd3dd54dc017b0a9167f85ce92f1711df7b0b9/src/ros_filter.cpp#L958

Eventually, we reach the loop condition check here:

https://github.com/cra-ros-pkg/robot_localization/blob/70cd3dd54dc017b0a9167f85ce92f1711df7b0b9/src/ros_filter.cpp#L1083

`more_params` is true, so we go back to the beginning of the loop.

We don't have `odom1` defined, so we'll hit this block:

https://github.com/cra-ros-pkg/robot_localization/blob/70cd3dd54dc017b0a9167f85ce92f1711df7b0b9/src/ros_filter.cpp#L961

...but `more_params` is still true. So we'll reach the loop condition check at the end and keep going. I am not currently seeing how this doesn't end in an infinite loop, but I'm assuming people are actually running this code, so I'm clearly missing something.

In the original ROS1 code, we are updating `moreParams` at every loop cycle:

https://github.com/cra-ros-pkg/robot_localization/blob/3bd285e6fc400447086490e833a6f0181284b4d8/src/ros_filter.cpp#L1034

In the new code, we set it to _true_ if it exists, but don't set it to _false_ if it doesn't.

Reply:
@ayrton04 is right, this `more_params` is missing a clear, I needed to add this to get things working

Reply:
Gah 🤦‍♂, ok thanks for taking another look, your def right.